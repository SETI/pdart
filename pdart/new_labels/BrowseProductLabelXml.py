"""A document template to create a label for a raw browse product."""
from pdart.xml.Pds4Version import *
from pdart.xml.Templates import *

make_label = interpret_document_template(
    """<?xml version="1.0" encoding="utf-8"?>
<?xml-model href="http://pds.nasa.gov/pds4/pds/v1/PDS4_PDS_%s.sch"
            schematypens="http://purl.oclc.org/dsdl/schematron"?>
<Product_Browse
   xmlns="http://pds.nasa.gov/pds4/pds/v1"
   xmlns:pds="http://pds.nasa.gov/pds4/pds/v1"
   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
   xsi:schemaLocation="http://pds.nasa.gov/pds4/pds/v1/PDS4_PDS_%s.xsd">
  <Identification_Area>
    <logical_identifier><NODE name="browse_lid" /></logical_identifier>
    <version_id>0.1</version_id>
    <title>This product contains a browse image of a <NODE name="suffix" /> \
image obtained the HST Observing Program <NODE name="proposal_id" />\
.</title>
    <information_model_version>%s</information_model_version>
    <product_class>Product_Browse</product_class>
    <Modification_History>
      <Modification_Detail>
        <modification_date>2016-04-20</modification_date>
        <version_id>0.1</version_id>
        <description>PDS4 version-in-development of the product</description>
      </Modification_Detail>
    </Modification_History>
  </Identification_Area>
  <Reference_List>
    <Internal_Reference>
      <lid_reference><NODE name="data_lid" /></lid_reference>
      <reference_type>browse_to_data</reference_type>
    </Internal_Reference>
  </Reference_List>
  <File_Area_Browse>
    <File>
      <file_name><NODE name="browse_file_name" /></file_name>
    </File>
    <Encoded_Image>
      <offset unit="byte">0</offset>
      <object_length unit="byte"><NODE name="object_length" /></object_length>
      <encoding_standard_id>JPEG</encoding_standard_id>
    </Encoded_Image>
  </File_Area_Browse>
</Product_Browse>""" %
    (PDS4_SHORT_VERSION, PDS4_SHORT_VERSION, INFORMATION_MODEL_VERSION))
# type: DocTemplate
"""
An interpreted document template to create a label for a RAW browse product.
"""
