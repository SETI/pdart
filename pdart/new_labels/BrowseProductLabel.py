"""
Functionality to create a label for a browse product containing browse
images.
"""
from typing import TYPE_CHECKING

from pdart.new_labels.BrowseProductLabelXml import make_label
from pdart.new_labels.Utils import lidvid_to_lid, lidvid_to_vid
from pdart.xml.Pretty import pretty_and_verify

if TYPE_CHECKING:
    from pdart.new_db.BundleDB import BundleDB


def make_browse_product_label(bundle_db,
                              browse_product_lidvid,
                              browse_file_basename,
                              verify):
    # type: (BundleDB, str, str, unicode, bool) -> unicode
    """
    Create the label text for the browse product having the given
    LIDVID using the bundle database.  If verify is True, verify the
    label against its XML and Schematron schemas.  Raise an exception
    if either fails.
    """
    fits_product_lidvid = bundle_db.get_product(
        browse_product_lidvid).fits_product_lidvid
    browse_file = bundle_db.get_file(browse_file_basename,
                                     browse_product_lidvid)

    browse_product = bundle_db.get_product(browse_product_lidvid)

    browse_collection_lidvid = browse_product.collection_lidvid
    browse_collection = bundle_db.get_collection(browse_collection_lidvid)

    bundle = bundle_db.get_bundle()

    label = make_label({
        'proposal_id': bundle.proposal_id,
        'suffix': browse_collection.suffix,
        'browse_lid': lidvid_to_lid(browse_product_lidvid),
        'browse_vid': lidvid_to_vid(browse_product_lidvid),
        'data_lidvid': fits_product_lidvid,
        'browse_file_name': browse_file_basename,
        'object_length': browse_file.byte_size
    }).toxml()

    return pretty_and_verify(label, verify)
