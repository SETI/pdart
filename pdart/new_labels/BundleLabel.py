"""Functionality to build a bundle label using a SQLite database."""
from typing import TYPE_CHECKING

from pdart.new_labels.BundleLabelXml \
    import make_bundle_entry_member, make_label
from pdart.new_labels.CitationInformation \
    import make_citation_information
from pdart.new_labels.Utils import lidvid_to_lid, lidvid_to_vid
from pdart.xml.Pretty import pretty_and_verify
from pdart.xml.Templates import combine_nodes_into_fragment

if TYPE_CHECKING:
    from Citation_Information import Citation_Information
    from pdart.new_db.BundleDB import BundleDB


def make_bundle_label(bundle_db, info, verify):
    # type: (BundleDB, Citation_Information, bool) -> unicode
    """
    Create the label text for the bundle in the bundle database using
    the database connection.  If verify is True, verify the label
    against its XML and Schematron schemas.  Raise an exception if
    either fails.
    """
    bundle = bundle_db.get_bundle()
    bundle_lid = lidvid_to_lid(bundle.lidvid)  # Only used in placeholder
    proposal_id = bundle.proposal_id
    reduced_collections = [
        make_bundle_entry_member({'collection_lidvid': collection.lidvid})
        for collection
        in bundle_db.get_bundle_collections(bundle.lidvid)]

    label = make_label({
        'bundle_lid': lidvid_to_lid(bundle.lidvid),
        'bundle_vid': lidvid_to_vid(bundle.lidvid),
        'proposal_id': str(proposal_id),
        'Citation_Information': make_citation_information(info),
        'Bundle_Member_Entries': combine_nodes_into_fragment(
            reduced_collections)
    }).toxml()

    return pretty_and_verify(label, verify)
