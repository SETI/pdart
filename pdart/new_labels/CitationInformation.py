"""
Functionality to create a ``<Citation_Information />`` XML element.
"""
from typing import TYPE_CHECKING

from Citation_Information import Citation_Information
from pdart.new_labels.Placeholders import known_placeholder, placeholder_year
from pdart.xml.Templates import interpret_template

if TYPE_CHECKING:
    from pdart.xml.Templates import NodeBuilder, NodeBuilderTemplate

_citation_information = interpret_template(
    """<Citation_Information>
<publication_year><NODE name="publication_year" /></publication_year>
<description><NODE name="description" /></description>
</Citation_Information>""")  # type: NodeBuilderTemplate
"""
An interpreted template to create a ``<Citation_Information
/>`` XML element.
"""


def make_citation_information(info):
    # type: (Citation_Information) -> NodeBuilder
    """
    Return a placeholder ``<Citation_Information />`` XML element.
    """
    return _citation_information({
        'description': info.description,
        'publication_year': info.publication_year
    })
