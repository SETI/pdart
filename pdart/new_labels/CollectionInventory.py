"""
Functionality to build a collection inventory using a SQLite database.
"""

from typing import TYPE_CHECKING

from pdart.new_labels.Utils import lidvid_to_lid

if TYPE_CHECKING:
    from pdart.new_db.BundleDB import BundleDB


def make_collection_inventory(bundle_db, collection_lidvid):
    # type: (BundleDB, str) -> unicode
    products = bundle_db.get_collection_products(
        collection_lidvid)
    inventory_lines = [u'P,%s\r\n' % lidvid_to_lid(product.lidvid)
                       for product in products]
    return ''.join(inventory_lines)
