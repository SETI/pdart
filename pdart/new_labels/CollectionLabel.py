"""
Functionality to build a collection label using a SQLite database.
"""

from typing import TYPE_CHECKING

from pdart.new_labels.CitationInformation \
    import make_placeholder_citation_information
from pdart.new_labels.CollectionLabelXml import make_label
from pdart.new_labels.Utils import lidvid_to_lid, lidvid_to_vid
from pdart.xml.Pretty import pretty_and_verify

if TYPE_CHECKING:
    from pdart.new_db.BundleDB import BundleDB


# TODO Should probably test document_collection independently.


def make_collection_label(bundle_db, collection_lidvid, verify):
    # type: (BundleDB, str, bool) -> unicode
    """
    Create the label text for the collection having this LIDVID using
    the bundle database.  If verify is True, verify the label against
    its XML and Schematron schemas.  Raise an exception if either
    fails.
    """
    collection_lid = lidvid_to_lid(collection_lidvid)
    collection_vid = lidvid_to_vid(collection_lidvid)
    collection = bundle_db.get_collection(collection_lidvid)
    suffix = collection.suffix
    proposal_id = bundle_db.get_bundle().proposal_id
    try:
        inventory_name = 'collection_%s.csv' % collection.prefix
    except:
        # Document collections won't have prefixes.
        inventory_name = 'collection.csv'

    label = make_label({
        'collection_lid': collection_lid,
        'collection_vid': collection_vid,
        'suffix': suffix,
        'proposal_id': str(proposal_id),
        'Citation_Information': make_placeholder_citation_information(
            collection_lid),
        'inventory_name': inventory_name
    }).toxml()

    return pretty_and_verify(label, verify)
