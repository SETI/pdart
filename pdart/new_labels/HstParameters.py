"""
Functionality to build an ``<hst:HST />`` XML element using a SQLite
database.
"""
from typing import TYPE_CHECKING

from pdart.new_labels.HstParametersXml import get_pc1_flag, \
    get_targeted_detector_id, get_wf2_flag, get_wf3_flag, get_wf4_flag, hst, \
    parameters_acs, parameters_general, parameters_wfc3, parameters_wfpc2
from pdart.new_labels.Placeholders import known_placeholder, placeholder, \
    placeholder_float, placeholder_int
from pdart.rules.Combinators import multiple_implementations

if TYPE_CHECKING:
    from typing import Any, Callable, Dict, List, Optional
    from pdart.xml.Templates import NodeBuilder


def get_repeat_exposure_count(product_id):
    # type: (unicode) -> unicode
    """
    Return a placeholder integer for the ``<repeat_exposure_count
    />`` XML element, noting the problem.
    """
    return placeholder_int(product_id, 'repeat_exposure_count')


def _get_subarray_flag(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    """
    Return placeholder text for the ``<subarray_flag />`` XML element,
    noting the problem.
    """
    if instrument != 'wfpc2':
        return card_dicts[0]['SUBARRAY']
    assert False

def _get_subarray_flag_placeholder(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    return placeholder(product_id, 'subarry_flag')

get_subarray_flag = multiple_implementations(
    'get_subarray_flag',
    _get_subarray_flag,
    _get_subarray_flag_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode, unicode], unicode]
"""
Return text for the ``<subarray_flag />`` XML element.
"""


##############################
# get_aperture_name
##############################


def _get_aperture_name(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    if instrument == 'wfpc2':
        try:
            return card_dicts[0]['APERTURE']
        except KeyError:
            return card_dicts[0]['APEROBJ']
    else:
        return card_dicts[0]['APERTURE']


def _get_aperture_name_placeholder(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    return placeholder(product_id, 'aperture_name')


get_aperture_name = multiple_implementations(
    'get_aperture_name',
    _get_aperture_name,
    _get_aperture_name_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode, unicode], unicode]
"""
Return text for the ``<aperture_name />`` XML element.
"""


##############################
# get_bandwidth
##############################


def _get_bandwidth(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    if instrument == 'wfpc2':
        bandwid = float(card_dicts[0]['BANDWID'])
        return str(bandwid * 1.0e-4)
    assert False


def _get_bandwidth_placeholder(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    return placeholder_float(product_id, 'bandwidth')


get_bandwidth = multiple_implementations(
    'get_bandwidth',
    _get_bandwidth,
    _get_bandwidth_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode, unicode], unicode]
"""
Return a float for the ``<bandwidth />`` XML element.
"""


##############################
# get_center_filter_wavelength
##############################

def _get_center_filter_wavelength(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    if instrument == 'wfpc2':
        centrwv = float(card_dicts[0]['CENTRWV'])
        return str(centrwv * 1.0e-4)
    else:
        raise Exception('Unhandled instrument %s' % instrument)


def _get_center_filter_wavelength_placeholder(card_dicts,
                                              instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    return placeholder_float(product_id, 'center_filter_wavelength')


get_center_filter_wavelength = multiple_implementations(
    'get_center_filter_wavelength',
    _get_center_filter_wavelength,
    _get_center_filter_wavelength_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode, unicode], unicode]
"""
Return a float for the ``<center_filter_wavelength />`` XML element.
"""


##############################
# get_detector_id
##############################

def _get_detector_id(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    detector = card_dicts[0]['DETECTOR']
    if instrument == 'wfpc2':
        if detector == '1':
            return 'PC1'
        else:
            return 'WF' + detector
    else:
        return detector


def _get_detector_id_placeholder(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    return placeholder(product_id, 'detector_id')


get_detector_id = multiple_implementations(
    'get_detector_id',
    _get_detector_id,
    _get_detector_id_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode, unicode], unicode]
"""
Return text for the ``<detector_id />`` XML element.
"""


##############################
# get_exposure_duration
##############################

def _get_exposure_duration(card_dicts, product_id):
    # type: (List[Dict[str, Any]], unicode) -> unicode
    return str(card_dicts[0]['EXPTIME'])


def _get_exposure_duration_placeholder(card_dicts, product_id):
    # type: (List[Dict[str, Any]], unicode) -> unicode
    return placeholder_float(product_id, 'exposure_duration')


get_exposure_duration = multiple_implementations(
    'get_exposure_duration',
    _get_exposure_duration,
    _get_exposure_duration_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode], unicode]
"""
Return a float for the ``<exposure_duration />`` XML element.
"""


##############################
# get_exposure_type
##############################

def _get_exposure_type(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], str, str) -> unicode
    if instrument == 'acs':
        try:
            return card_dicts[0]['EXPFLAG']
        except KeyError:
            return 'UNK'
    else:
        return card_dicts[0]['EXPFLAG']


def _get_exposure_type_placeholder(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], str, str) -> unicode
    return placeholder(product_id, 'exposure_type')


get_exposure_type = multiple_implementations(
    'get_exposure_type',
    _get_exposure_type,
    _get_exposure_type_placeholder
)  # type: Callable[[List[Dict[str, Any]], str, str], unicode]
"""
Return text for the ``<exposure_type />`` XML element.
"""


##############################
# get_filter_name
##############################

def _get_filter_name(card_dicts, instrument, product_name):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    if instrument == 'wfpc2':
        filtnam1 = card_dicts[0]['FILTNAM1'].strip()
        filtnam2 = card_dicts[0]['FILTNAM2'].strip()
        if filtnam1 == '':
            return filtnam2
        elif filtnam2 == '':
            return filtnam1
        else:
            return '%s+%s' % (filtnam1, filtnam2)
    elif instrument == 'acs':
        filter1 = card_dicts[0]['FILTER1'].strip()
        filter2 = card_dicts[0]['FILTER2'].strip()
        if filter1.startswith('CLEAR'):
            if filter2.startswith('CLEAR'):
                return 'CLEAR'
            else:
                return filter2
        else:
            if filter2.startswith('CLEAR'):
                return filter1
            else:
                return '%s+%s' % (filter1, filter2)
    else:
        assert (instrument == 'wfc3')
        return card_dicts[0]['FILTER']


def _get_filter_name_placeholder(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    return placeholder(product_id, 'filter_name')


get_filter_name = multiple_implementations(
    'get_filter_name',
    _get_filter_name,
    _get_filter_name_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode, unicode], unicode]
"""
Return text for the ``<filter_name />`` XML element.
"""


##############################
# get_fine_guidance_system_lock_type
##############################

def _get_fine_guidance_system_lock_type(card_dicts, product_id):
    try:
        return card_dicts[0]['FGSLOCK']
    except KeyError:
        return 'UNK'


def _get_fine_guidance_system_lock_type_placeholder(card_dicts, product_id):
    return placeholder(product_id, 'fine_guidance_system_lock_type')


get_fine_guidance_system_lock_type = multiple_implementations(
    'get_fine_guidance_system_lock_type',
    _get_fine_guidance_system_lock_type,
    _get_fine_guidance_system_lock_type_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode], unicode]
"""
Return text for the ``<fine_guidance_system_lock_type />`` XML element.
"""


##############################
# get_gain_mode_id
##############################


def _get_gain_mode_id(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    try:
        atodgain = card_dicts[0]['ATODGAIN']
    except KeyError:
        return 'N/A'
    if instrument == 'acs':
        return str(atodgain)
    elif instrument == 'wfpc2':
        return 'A2D' + str(int(atodgain))
    assert False


def _get_gain_mode_id_placeholder(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    return placeholder(product_id, 'gain_mode_id')


get_gain_mode_id = multiple_implementations(
    'get_gain_mode_id',
    _get_gain_mode_id,
    _get_gain_mode_id_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode, unicode], unicode]
"""
Return text for the ``<gain_mode_id />`` XML element.
"""


##############################
# get_gyroscope_mode
##############################

def _get_gyroscope_mode(card_dicts, product_id):
    return card_dicts[0]['GYROMODE']

def _get_gyroscope_mode_placeholder(card_dicts, product_id):
    return placeholder(product_id, 'gyroscope_mode')

get_gyroscope_mode  = multiple_implementations(
    'get_gyroscope_mode',
    _get_gyroscope_mode,
    _get_gyroscope_mode_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode], unicode]


##############################
# get_hst_pi_name
##############################

def _get_hst_pi_name(card_dicts, product_id):
    # type: (List[Dict[str, Any]], unicode) -> unicode
    pr_inv_l = card_dicts[0]['PR_INV_L']
    pr_inv_f = card_dicts[0]['PR_INV_F']
    try:
        pr_inv_m = card_dicts[0]['PR_INV_M']
        return '%s, %s %s' % (pr_inv_l, pr_inv_f, pr_inv_m)
    except KeyError:
        return '%s, %s' % (pr_inv_l, pr_inv_f)


def _get_hst_pi_name_placeholder(card_dicts, product_id):
    # type: (List[Dict[str, Any]], unicode) -> unicode
    return placeholder(product_id, 'hst_pi_name')


get_hst_pi_name = multiple_implementations(
    'get_hst_pi_name',
    _get_hst_pi_name,
    _get_hst_pi_name_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode], unicode]
"""
Return text for the ``<hst_pi_name />`` XML element.
"""


##############################
# get_hst_proposal_id
##############################

def _get_hst_proposal_id(card_dicts, product_id):
    # type: (List[Dict[str, Any]], unicode) -> unicode
    return str(card_dicts[0]['PROPOSID'])


def _get_hst_proposal_id_placeholder(card_dicts, product_id):
    # type: (List[Dict[str, Any]], unicode) -> unicode
    return placeholder_int(product_id, 'hst_proposal_id')


get_hst_proposal_id = multiple_implementations(
    'get_hst_proposal_id',
    _get_hst_proposal_id,
    _get_hst_proposal_id_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode], unicode]
"""
Return text for the ``<hst_proposal_id />`` XML element.
"""


##############################
# get_hst_target_name
##############################

def _get_hst_target_name(card_dicts, product_id):
    # type: (List[Dict[str, Any]], unicode) -> unicode
    return card_dicts[0]['TARGNAME']


def _get_hst_target_name_placeholder(card_dicts, product_id):
    # type: (List[Dict[str, Any]], unicode) -> unicode
    return placeholder(product_id, 'hst_target_name')


get_hst_target_name = multiple_implementations(
    'get_hst_target_name',
    _get_hst_target_name,
    _get_hst_target_name_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode], unicode]
"""
Return text for the ``<hst_target_name />`` XML element.
"""


##############################
# get_instrument_mode_id
##############################

def _get_instrument_mode_id(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    if instrument == 'acs':
        try:
            return card_dicts[0]['OBSMODE']
        except KeyError:
            return 'UNK'
    if instrument == 'wfpc2':
        return card_dicts[0]['MODE']
    else:
        assert (instrument == 'wfc3')
        return card_dicts[0]['OBSMODE']


def _get_instrument_mode_id_placeholder(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    return placeholder(product_id, 'instrument_mode_id')


get_instrument_mode_id = multiple_implementations(
    'get_instrument_mode_id',
    _get_instrument_mode_id,
    _get_instrument_mode_id_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode, unicode], unicode]
"""
Return text for the ``<instrument_mode_id />`` XML element.
"""


##############################
# get_observation_type
##############################

def _get_observation_type(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    if instrument != 'wfpc2':
        return card_dicts[0]['OBSTYPE']
    else:
        raise Exception('Unhandled instrument %s' % instrument)


def _get_observation_type_placeholder(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], unicode, unicode) -> unicode
    return placeholder(product_id, 'observation_type')


get_observation_type = multiple_implementations(
    'get_observation_type',
    _get_observation_type,
    _get_observation_type_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode, unicode], unicode]
"""
Return text for the ``<observation_type />`` XML element.
"""

##############################
# get_stsci_group_id: NOTE the name is going to change
##############################

def _get_stsci_group_id(card_dicts, product_id):
    try:
        asn_id = card_dicts[0]['ASN_ID']
        if asn_id == 'NONE':
            return card_dicts[0]['ROOTNAME']
        else:
            return asn_id
    except KeyError:
        return card_dicts[0]['ROOTNAME']


def _get_stsci_group_id_placeholder(card_dicts, product_id):
    # type: (List[Dict[str, Any]], unicode) -> unicode
    return placeholder(product_id, 'stsci_group_id')

get_stsci_group_id = multiple_implementations(
    'get_stsci_group_id',
    _get_stsci_group_id,
    _get_stsci_group_id_placeholder
)  # type: Callable[[List[Dict[str, Any]], unicode], unicode]
"""
Return text for the ``<stsci_group_id />`` XML element.
"""

##############################

def get_hst_parameters(card_dicts, instrument, product_id):
    # type: (List[Dict[str, Any]], str, str) -> NodeBuilder
    """Return an ``<hst:HST />`` XML element."""
    d = {'stsci_group_id': get_stsci_group_id(card_dicts, product_id),
         'hst_proposal_id': get_hst_proposal_id(card_dicts, product_id),
         'hst_pi_name': get_hst_pi_name(card_dicts, product_id),
         'hst_target_name': get_hst_target_name(card_dicts, product_id),
         'aperture_name': get_aperture_name(card_dicts, instrument,
                                            product_id),
         'exposure_duration': get_exposure_duration(card_dicts, product_id),
         'exposure_type':
             get_exposure_type(card_dicts, instrument, product_id),
         'filter_name': get_filter_name(card_dicts, instrument, product_id),
         'fine_guidance_system_lock_type':
             get_fine_guidance_system_lock_type(card_dicts, product_id),
         'gyroscope_mode': get_gyroscope_mode(card_dicts, product_id),
         'instrument_mode_id': get_instrument_mode_id(card_dicts,
                                                      instrument,
                                                      product_id),
         'moving_target_flag': 'true'}

    if instrument == 'acs':
        parameters_instrument = parameters_acs(
            {'detector_id': get_detector_id(card_dicts, instrument,
                                            product_id),
             'gain_mode_id': get_gain_mode_id(card_dicts, instrument,
                                              product_id),
             'observation_type':
                 get_observation_type(card_dicts, instrument, product_id),
             'repeat_exposure_count': get_repeat_exposure_count(product_id),
             'subarray_flag':
                 get_subarray_flag(card_dicts, instrument, product_id)})
    elif instrument == 'wfpc2':
        parameters_instrument = parameters_wfpc2(
            {'bandwidth': get_bandwidth(card_dicts, instrument, product_id),
             'center_filter_wavelength':
                 get_center_filter_wavelength(card_dicts,
                                              instrument,
                                              product_id),
             'targeted_detector_id':
                 get_targeted_detector_id(card_dicts[0]['APERTURE']),
             'gain_mode_id': get_gain_mode_id(card_dicts, instrument,
                                              product_id),
             'pc1_flag': get_pc1_flag(product_id, instrument),
             'wf2_flag': get_wf2_flag(product_id, instrument),
             'wf3_flag': get_wf3_flag(product_id, instrument),
             'wf4_flag': get_wf4_flag(product_id, instrument)})
    elif instrument == 'wfc3':
        parameters_instrument = parameters_wfc3(
            {'detector_id': get_detector_id(card_dicts,
                                            instrument,
                                            product_id),
             'observation_type':
                 get_observation_type(card_dicts, instrument, product_id),
             'repeat_exposure_count':
                 get_repeat_exposure_count(product_id),
             'subarray_flag':
                 get_subarray_flag(card_dicts, instrument, product_id)})
    else:
        assert False, 'Bad instrument value: %s' % instrument

    return hst({
        'parameters_general': parameters_general(d),
        'parameters_instrument': parameters_instrument})
