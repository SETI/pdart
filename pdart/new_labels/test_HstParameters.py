import unittest

from fs.path import basename, join

from pdart.new_db.BundleDB import create_bundle_db_in_memory
from pdart.new_db.FitsFileDB import card_dictionaries, populate_from_fits_file
from pdart.new_db.SqlAlchTables import File, FitsFile
from pdart.new_labels.HstParameters import *
from pdart.xml.Pretty import pretty_print


class Test_HstParameters(unittest.TestCase):
    # TODO Write cases for other two instruments

    # TODO get_gain_mode_id() is failing on this input file.  Investigate why.

    def test_get_acs_parameters(self):
        # type: () -> None
        db = create_bundle_db_in_memory()
        db.create_tables()
        archive = '/Users/spaceman/Desktop/Archive'

        fits_product_lidvid = \
            'urn:nasa:pds:hst_13012:data_acs_raw:jbz504eoq_raw::2'
        os_filepath = join(
            archive,
            'hst_13012/data_acs_raw/visit_04/jbz504eoq_raw.fits')

        populate_from_fits_file(db,
                                os_filepath,
                                fits_product_lidvid)

        file_basename = basename(os_filepath)

        fits_file = db.session.query(FitsFile).filter(
            File.product_lidvid == fits_product_lidvid).filter(
            File.basename == file_basename).one()

        hdu_count = fits_file.hdu_count

        card_dicts = card_dictionaries(db.session,
                                       fits_product_lidvid,
                                       hdu_count)

        nb = get_hst_parameters(card_dicts, u'acs', fits_product_lidvid)
        doc = xml.dom.getDOMImplementation().createDocument(None, None, None)
        str = nb(doc).toxml()
        str = pretty_print(str)
        print str
        self.assertEqual(_expected, str)


_expected = '\n'.join([
    '<?xml version="1.0"?>',
    '<hst:HST>',
    '  <hst:Parameters_General>',
    '    <hst:stsci_group_id>### placeholder for stsci_group_id '
    '###</hst:stsci_group_id>',
    '    <hst:hst_proposal_id>13012</hst:hst_proposal_id>',
    '    <hst:hst_pi_name>Lamy, Laurent </hst:hst_pi_name>',
    '    <hst:hst_target_name>URANUS</hst:hst_target_name>',
    '    <hst:aperture_type>SBC</hst:aperture_type>',
    '    <hst:exposure_duration>270.0</hst:exposure_duration>',
    '    <hst:exposure_type>NORMAL</hst:exposure_type>',
    '    <hst:filter_name>F165LP+N/A</hst:filter_name>',
    '    <hst:fine_guidance_system_lock_type>FINE</hst'
    ':fine_guidance_system_lock_type>',
    '    <hst:gyroscope_mode>### placeholder for gyroscope_mode '
    '###</hst:gyroscope_mode>',
    '    <hst:instrument_mode_id>ACCUM</hst:instrument_mode_id>',
    '    <hst:moving_target_flag>true</hst:moving_target_flag>',
    '  </hst:Parameters_General>',
    '  <hst:Parameters_ACS>',
    '    <hst:detector_id>SBC</hst:detector_id>',
    '    <hst:gain_mode_id>### placeholder for gain_mode_id '
    '###</hst:gain_mode_id>',
    '    <hst:observation_type>IMAGING</hst:observation_type>',
    '    <hst:repeat_exposure_count>0</hst:repeat_exposure_count>',
    '    <hst:subarray_flag>### placeholder for subarray_flag ###'
    '</hst:subarray_flag>',
    '  </hst:Parameters_ACS>',
    '</hst:HST>',
    ''])  # type: str
# I'm jumping through hoops here just to get PEP8 compliance.
