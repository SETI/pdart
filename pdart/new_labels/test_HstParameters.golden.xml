<?xml version="1.0"?>
<hst:HST>
  <hst:Parameters_General>
    <hst:stsci_group_id>### placeholder for stsci_group_id ###</hst:stsci_group_id>
    <hst:hst_proposal_id>13012</hst:hst_proposal_id>
    <hst:hst_pi_name>Lamy, Laurent </hst:hst_pi_name>
    <hst:hst_target_name>URANUS</hst:hst_target_name>
    <hst:aperture_type>SBC</hst:aperture_type>
    <hst:exposure_duration>270.0</hst:exposure_duration>
    <hst:exposure_type>NORMAL</hst:exposure_type>
    <hst:filter_name>F165LP+N/A</hst:filter_name>
    <hst:fine_guidance_system_lock_type>FINE</hst:fine_guidance_system_lock_type>
    <hst:gyroscope_mode>### placeholder for gyroscope_mode ###</hst:gyroscope_mode>
    <hst:instrument_mode_id>ACCUM</hst:instrument_mode_id>
    <hst:moving_target_flag>true</hst:moving_target_flag>
  </hst:Parameters_General>
  <hst:Parameters_ACS>
    <hst:detector_id>SBC</hst:detector_id>
    <hst:gain_mode_id>### placeholder for gain_mode_id ###</hst:gain_mode_id>
    <hst:observation_type>IMAGING</hst:observation_type>
    <hst:repeat_exposure_count>0</hst:repeat_exposure_count>
    <hst:subarray_flag>### placeholder for subarray_flag ###</hst:subarray_flag>
  </hst:Parameters_ACS>
</hst:HST>
